src/NotaryService/NotaryService.cs

```diff
@@ -0,0 +1,401 @@
+                return;
+            }
+            witnessInfos = new WitnessInfo[tx.Signers.Length];
+            for (int i = 0; i < tx.Witnesses.Length; i++)
+            {
+                if (tx.Witnesses[i].VerificationScript.Length == 0)
+                {
+                    witnessInfos[i] = new WitnessInfo()
+                    {
+                        Typ = RequestType.Contract,
+                        NSigsLeft = 0
+                    };
+                    continue;
+                }
+                if (!tx.Signers[i].Account.Equals(tx.Witnesses[i].VerificationScript.ToScriptHash()))
+                {
+                    validationErr = $"transaction should have valid verification script for signer {i}";
+                    witnessInfos = null;
+                    return;
+                }
+                if (tx.Witnesses[i].InvocationScript.Length != 0)
+                {
+                    if (tx.Witnesses[i].InvocationScript.Length != 66 || (tx.Witnesses[i].InvocationScript[0] != (byte)OpCode.PUSHDATA1 && tx.Witnesses[i].InvocationScript[1] != 64))
+                    {
+                        validationErr = "multisignature invocation script should have length = 66 and be of the form[PUSHDATA1, 64, signatureBytes...]";
+                        witnessInfos = null;
+                        return;
+                    }
+                }
+                if (tx.Witnesses[i].VerificationScript.IsMultiSigContract(out int nSigs, out ECPoint[] pubs))
+                {
+                    witnessInfos[i] = new WitnessInfo()
+                    {
+                        Typ = RequestType.MultiSignature,
+                        NSigsLeft = (byte)nSigs,
+                        Pubs = pubs,
+                    };
+                    nKeysActual += pubs.Length;
+                    continue;
+                }
+                if (tx.Witnesses[i].VerificationScript.IsSignatureContract())
+                {
+                    witnessInfos[i] = new WitnessInfo()
+                    {
+                        Typ = RequestType.Signature,
+                        NSigsLeft = 1,
+                        Pubs = new ECPoint[pubs.Length],
+                    };
+                    nKeysActual++;
+                    continue;
+                }
+                validationErr = $"witness {i}: unable to detect witness type, only sig/multisig/contract are supported";
+                witnessInfos = null;
+                return;
+            }
+            if (nKeysActual != nKeysExpected)
+            {
+                validationErr = $"expected and actual NKeys mismatch: {nKeysExpected} vs {nKeysActual}";
+                witnessInfos = null;
+                return;
+            }
+            return;
+        }
+
+        private bool VerifyWitness(IVerifiable verifiable, ProtocolSettings settings, DataCache snapshot, UInt160 hash, Witness witness, long gas, out long fee)
```

Comment:
It looks to me this duplicates some code from core Neo, do we need this?

Reply:
Got it